File: sdk/python/easynet_sdk/ability_descriptor.py
"""AbilityDescriptor projection through runtime catalog facts."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Mapping, Protocol, cast

from .errors import ErrorCode, RetryHint, SDKError
from .runtime_ability import RuntimeAbilityClient, RuntimeCallContext

__all__ = [
    "AbilityDescriptorClient",
    "AbilityDescriptorGetRequest",
    "AbilityDescriptorHints",
    "AbilityDescriptorListRequest",
    "AbilityDescriptorPage",
    "AbilityDescriptorProjection",
    "AbilityDescriptorProvider",
    "AbilityDescriptorRef",
    "RuntimeAbilityDescriptorProvider",
    "parse_ability_descriptor_ref",
    "project_ability_descriptor",
]

_RUNTIME_ABILITY_DESCRIPTOR_LIST_ROUTE = "meta.list_abilities"


@dataclass(frozen=True)
class AbilityDescriptorHints:
    """Advisory transport and behavior hints from the governed descriptor."""

    read_only: bool = False
    destructive: bool = False
    idempotent: bool = False
    streaming_only: bool = False
    bidi_only: bool = False


@dataclass(frozen=True)
class AbilityDescriptorProjection:
    """Generic SDK read model for one runtime AbilityDescriptor row."""

    ability_ura: str = ""
    descriptor_ref: str = ""
    name: str = ""
    owner_ura: str = ""
    version: str = ""
    schema_hash: str = ""
    descriptor_hash: str = ""
    call_mode: str = ""
    class_: str = ""
    receipt_semantics: Mapping[str, object] = field(default_factory=dict)
    visibility: str = ""
    source: str = ""
    description: str = ""
    hints: AbilityDescriptorHints = field(default_factory=AbilityDescriptorHints)
    schema_summary: Mapping[str, object] = field(default_factory=dict)
    input_schema: Mapping[str, object] = field(default_factory=dict)
    metadata: Mapping[str, object] = field(default_factory=dict)


@dataclass(frozen=True)
class AbilityDescriptorRef:
    """Descriptor identity split into ability URA and descriptor version."""

    raw: str
    ability_ura: str
    version: str


@dataclass(frozen=True)
class AbilityDescriptorListRequest:
    """Request runtime descriptor catalog rows with generic filters."""

    call: RuntimeCallContext
    scope: str = ""
    owner_ura: str = ""
    ability_ura: str = ""


@dataclass(frozen=True)
class AbilityDescriptorGetRequest:
    """Resolve one canonical AbilityDescriptor row from the runtime catalog."""

    call: RuntimeCallContext
    ability_ura: str
    descriptor_version: str = ""
    call_mode: str = ""
    scope: str = ""


@dataclass(frozen=True)
class AbilityDescriptorPage:
    descriptors: tuple[AbilityDescriptorProjection, ...]


class AbilityDescriptorProvider(Protocol):
    def list(self, request: AbilityDescriptorListRequest) -> AbilityDescriptorPage: ...

    def get(self, request: AbilityDescriptorGetRequest) -> AbilityDescriptorProjection: ...


class _AddressingProjector(Protocol):
    def project_descriptor_ref(self, value: str) -> Any: ...


class AbilityDescriptorClient:
    """Stable product-neutral descriptor facade."""

    def __init__(self, provider: AbilityDescriptorProvider) -> None:
        if provider is None:
            raise _invalid_descriptor("AbilityDescriptor provider is required")
        self._provider = provider

    def list(self, request: AbilityDescriptorListRequest) -> AbilityDescriptorPage:
        return self._provider.list(request)

    def get(self, request: AbilityDescriptorGetRequest) -> AbilityDescriptorProjection:
        return self._provider.get(request)


class RuntimeAbilityDescriptorProvider:
    """Provider-backed descriptor catalog over the canonical runtime route."""

    def __init__(self, ability: RuntimeAbilityClient) -> None:
        if ability is None:
            raise _invalid_descriptor("runtime ability client is required")
        self._ability = ability

    def list(self, request: AbilityDescriptorListRequest) -> AbilityDescriptorPage:
        if not isinstance(request, AbilityDescriptorListRequest):
            raise _invalid_descriptor("AbilityDescriptorListRequest is required")
        args: dict[str, object] = {}
        if request.scope.strip():
            args["scope"] = request.scope.strip()
        if request.owner_ura.strip():
            args["owner_ura"] = request.owner_ura.strip()
        if request.ability_ura.strip():
            args["ability_ura"] = request.ability_ura.strip()
        output = self._ability._invoke_catalogue_read(
            request.call, _RUNTIME_ABILITY_DESCRIPTOR_LIST_ROUTE, args
        )
        raw_abilities = output.get("abilities")
        if not isinstance(raw_abilities, list):
            raise _invalid_descriptor(
                "runtime descriptor catalog output must include descriptor rows"
            )
        descriptors: list[AbilityDescriptorProjection] = []
        for index, raw in enumerate(raw_abilities):
            if not isinstance(raw, Mapping):
                raise _invalid_descriptor(
                    f"ability descriptor row {index} must be an object"
                )
            projection = _project_runtime_ability_descriptor(raw, index)
            if (
                not projection.ability_ura
                or not projection.owner_ura
                or not projection.name
                or not projection.version
            ):
                raise _invalid_descriptor(
                    f"ability descriptor row {index} is missing identity fields"
                )
            descriptors.append(projection)
        return AbilityDescriptorPage(tuple(descriptors))

    def get(self, request: AbilityDescriptorGetRequest) -> AbilityDescriptorProjection:
        if not isinstance(request, AbilityDescriptorGetRequest):
            raise _invalid_descriptor("AbilityDescriptorGetRequest is required")
        ability_ura = _required_text(request.ability_ura, "ability_ura")
        page = self.list(
            AbilityDescriptorListRequest(
                call=request.call,
                scope=request.scope,
                ability_ura=ability_ura,
            )
        )
        version = request.descriptor_version.strip()
        call_mode = request.call_mode.strip()
        matches: list[AbilityDescriptorProjection] = []
        for descriptor in page.descriptors:
            if descriptor.ability_ura != ability_ura:
                raise _invalid_descriptor(
                    "runtime returned descriptor outside requested ability_ura"
                )
            if version and descriptor.version != version:
                continue
            if call_mode and descriptor.call_mode != call_mode:
                continue
            matches.append(descriptor)
        if not matches:
            raise _not_found(ability_ura)
        if len(matches) > 1:
            raise _invalid_descriptor(
                "ability descriptor selection is ambiguous; specify descriptor_version or call_mode"
            )
        return matches[0]


def parse_ability_descriptor_ref(
    raw: str,
    *,
    addressing: object | None = None,
    library_path: str | None = None,
    control_path: str = "",
) -> AbilityDescriptorRef:
    """Project an AbilityDescriptorRef through the Axon-delegated SDK facade.

    Grammar ownership stays in the Identity/Addressing profile. Callers may
    inject an AddressingClient-like object for tests or reuse; otherwise the
    default SDK environment opens the configured C ABI facade.
    """

    try:
        if addressing is not None:
            projection = cast(_AddressingProjector, addressing).project_descriptor_ref(raw)
        else:
            from .axon_addressing import project_descriptor_ref

            projection = project_descriptor_ref(raw)
    except SDKError:
        raise
    except Exception as exc:
        raise _invalid_descriptor_ref("descriptor_ref projection failed", exc) from exc
    descriptor_ref = getattr(projection, "descriptor_ref", "")
    ability_ura = getattr(projection, "ability_ura", "")
    version = getattr(projection, "descriptor_version", "")
    if not descriptor_ref or not ability_ura or not version:
        raise _invalid_descriptor_ref("descriptor_ref projection is incomplete")
    return AbilityDescriptorRef(
        raw=descriptor_ref,
        ability_ura=ability_ura,
        version=version,
    )


def project_ability_descriptor(raw: Mapping[str, object]) -> AbilityDescriptorProjection:
    """Project one runtime descriptor row without deriving governed facts."""

    hints = _mapping(raw.get("hints"))
    schema = _mapping(raw.get("schema_summary"))
    return AbilityDescriptorProjection(
        ability_ura=_text(raw.get("ability_ura")),
        descriptor_ref=_text(raw.get("descriptor_ref")),
        name=_text(raw.get("name")),
        owner_ura=_text(raw.get("owner_ura")),
        version=_text(raw.get("descriptor_version")),
        schema_hash=_text(raw.get("schema_hash")),
        descriptor_hash=_text(raw.get("descriptor_hash")),
        call_mode=_text(raw.get("call_mode")),
        class_=_text(raw.get("class")),
        receipt_semantics=_mapping(raw.get("receipt_semantics")),
        visibility=_text(raw.get("visibility")),
        source=_text(raw.get("source")),
        description=_text(raw.get("description")),
        hints=AbilityDescriptorHints(
            read_only=_bool(hints.get("read_only")),
            destructive=_bool(hints.get("destructive")),
            idempotent=_bool(hints.get("idempotent")),
            streaming_only=_bool(hints.get("streaming_only")),
            bidi_only=_bool(hints.get("bidi_only")),
        ),
        schema_summary=schema,
        input_schema=_mapping(raw.get("input_schema")),
        metadata=_mapping(raw.get("metadata")),
    )


def _project_runtime_ability_descriptor(
    raw: Mapping[str, object], index: int
) -> AbilityDescriptorProjection:
    hints = _optional_mapping(raw, "hints", index)
    schema_summary = _optional_mapping(raw, "schema_summary", index)
    return AbilityDescriptorProjection(
        ability_ura=_required_descriptor_text(raw, "ability_ura", index),
        descriptor_ref=_required_descriptor_text(raw, "descriptor_ref", index),
        name=_required_descriptor_text(raw, "name", index),
        owner_ura=_required_descriptor_text(raw, "owner_ura", index),
        version=_required_descriptor_text(raw, "descriptor_version", index),
        schema_hash=_optional_descriptor_text(raw, "schema_hash", index),
        descriptor_hash=_optional_descriptor_text(raw, "descriptor_hash", index),
        call_mode=_optional_descriptor_text(raw, "call_mode", index),
        class_=_optional_descriptor_text(raw, "class", index),
        receipt_semantics=_optional_mapping(raw, "receipt_semantics", index),
        visibility=_optional_descriptor_text(raw, "visibility", index),
        source=_optional_descriptor_text(raw, "source", index),
        description=_optional_descriptor_text(raw, "description", index),
        hints=AbilityDescriptorHints(
            read_only=_optional_descriptor_bool(hints, "hints.read_only", index),
            destructive=_optional_descriptor_bool(hints, "hints.destructive", index),
            idempotent=_optional_descriptor_bool(hints, "hints.idempotent", index),
            streaming_only=_optional_descriptor_bool(
                hints, "hints.streaming_only", index
            ),
            bidi_only=_optional_descriptor_bool(hints, "hints.bidi_only", index),
        ),
        schema_summary=schema_summary,
        input_schema=_optional_mapping(raw, "input_schema", index),
        metadata=_optional_mapping(raw, "metadata", index),
    )


def _invalid_descriptor_ref(
    message: str, cause: BaseException | None = None
) -> SDKError:
    return SDKError(
        code=ErrorCode.INVALID_ARGUMENT,
        stage="descriptor_ref",
        retry=RetryHint.NEVER,
        retryable=False,
        message=message,
        cause=cause,
    )


def _text(value: object) -> str:
    return value.strip() if isinstance(value, str) else ""


def _bool(value: object) -> bool:
    return isinstance(value, bool) and value


def _mapping(value: object) -> dict[str, object]:
    return dict(value) if isinstance(value, Mapping) else {}


def _required_descriptor_text(
    raw: Mapping[str, object], field_name: str, index: int
) -> str:
    value = raw.get(field_name)
    if value is None:
        return ""
    if not isinstance(value, str):
        raise _invalid_descriptor(
            f"ability descriptor row {index} field {field_name} must be a string"
        )
    return value.strip()


def _optional_descriptor_text(
    raw: Mapping[str, object], field_name: str, index: int
) -> str:
    value = raw.get(field_name)
    if value is None:
        return ""
    if not isinstance(value, str):
        raise _invalid_descriptor(
            f"ability descriptor row {index} field {field_name} must be a string"
        )
    return value.strip()


def _optional_descriptor_bool(
    raw: Mapping[str, object], field_name: str, index: int
) -> bool:
    key = field_name.rsplit(".", 1)[-1]
    value = raw.get(key)
    if value is None:
        return False
    if not isinstance(value, bool):
        raise _invalid_descriptor(
            f"ability descriptor row {index} field {field_name} must be a boolean"
        )
    return value


def _optional_mapping(
    raw: Mapping[str, object], field_name: str, index: int
) -> dict[str, object]:
    value = raw.get(field_name)
    if value is None:
        return {}
    if not isinstance(value, Mapping):
        raise _invalid_descriptor(
            f"ability descriptor row {index} field {field_name} must be an object"
        )
    return dict(value)


def _required_text(value: object, field_name: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise _invalid_descriptor(f"{field_name} is required")
    return value.strip()


def _invalid_descriptor(
    message: str, cause: BaseException | None = None
) -> SDKError:
    return SDKError(
        code=ErrorCode.INVALID_ARGUMENT,
        stage="ability_descriptor",
        retry=RetryHint.NEVER,
        retryable=False,
        message=message,
        cause=cause,
    )


def _not_found(ability_ura: str) -> SDKError:
    return SDKError(
        code=ErrorCode.DESCRIPTOR_NOT_FOUND,
        stage="ability_descriptor",
        retry=RetryHint.NEVER,
        retryable=False,
        message="ability descriptor not found",
        details={"ability_ura": ability_ura},
    )

File: sdk/go/ability_descriptor.go
package easynet

import (
	"context"
	"fmt"
	"strings"
)

const runtimeAbilityDescriptorListRoute = "meta.list_abilities"

// AbilityDescriptorHints mirrors descriptor tool-annotation booleans that
// consumers may render without parsing ability names.
type AbilityDescriptorHints struct {
	ReadOnly      bool `json:"read_only"`
	Destructive   bool `json:"destructive"`
	Idempotent    bool `json:"idempotent"`
	StreamingOnly bool `json:"streaming_only"`
	BidiOnly      bool `json:"bidi_only"`
}

// AbilityDescriptorProjection is the SDK read-model projection of one
// advertised AbilityDescriptor. Product-specific display conventions remain in
// Metadata; this type owns only generic runtime descriptor fields.
type AbilityDescriptorProjection struct {
	AbilityURA       string
	DescriptorRef    string
	Name             string
	OwnerURA         string
	Version          string
	SchemaHash       string
	DescriptorHash   string
	CallMode         string
	Class            string
	ReceiptSemantics map[string]any
	Visibility       string
	Source           string
	Description      string
	Hints            AbilityDescriptorHints
	SchemaSummary    map[string]any
	InputSchema      map[string]any
	Metadata         map[string]any
}

// AbilityDescriptorRef is the SDK DTO projection of a descriptor identity.
// DescriptorRef grammar and canonicalization are owned by Axon behind the
// runtime identity profile boundary; this value object only carries the
// projected fields.
type AbilityDescriptorRef struct {
	Raw            string
	AbilityURA     string
	Version        string
	DescriptorHash string
	Action         string
}

// AbilityDescriptorListRequest asks the runtime catalog for
// generic AbilityDescriptor rows. The SDK exposes product-neutral filter names
// and lowers them to the provider catalog argument fields.
type AbilityDescriptorListRequest struct {
	Call       RuntimeCallContext `json:"call"`
	Scope      string             `json:"scope,omitempty"`
	OwnerURA   string             `json:"owner_ura,omitempty"`
	AbilityURA string             `json:"ability_ura,omitempty"`
}

type AbilityDescriptorGetRequest struct {
	Call              RuntimeCallContext `json:"call"`
	AbilityURA        string             `json:"ability_ura"`
	DescriptorVersion string             `json:"descriptor_version,omitempty"`
	CallMode          string             `json:"call_mode,omitempty"`
	Scope             string             `json:"scope,omitempty"`
}

type AbilityDescriptorPage struct {
	Descriptors []AbilityDescriptorProjection `json:"descriptors"`
}

// AbilityDescriptorProvider is the product-neutral catalog provider seam.
// Implementations must use runtime facts instead of name-derived governance.
type AbilityDescriptorProvider interface {
	List(context.Context, AbilityDescriptorListRequest) (AbilityDescriptorPage, error)
	Get(context.Context, AbilityDescriptorGetRequest) (AbilityDescriptorProjection, error)
}

type AbilityDescriptorClient struct {
	provider AbilityDescriptorProvider
}

func NewAbilityDescriptorClient(provider AbilityDescriptorProvider) (*AbilityDescriptorClient, error) {
	if provider == nil {
		return nil, invalidAbilityDescriptor("AbilityDescriptor provider is required", nil)
	}
	return &AbilityDescriptorClient{provider: provider}, nil
}

func (c *AbilityDescriptorClient) List(ctx context.Context, request AbilityDescriptorListRequest) (AbilityDescriptorPage, error) {
	if c == nil || c.provider == nil {
		return AbilityDescriptorPage{}, invalidAbilityDescriptor("AbilityDescriptor client is not initialized", nil)
	}
	return c.provider.List(ctx, request)
}

func (c *AbilityDescriptorClient) Get(ctx context.Context, request AbilityDescriptorGetRequest) (AbilityDescriptorProjection, error) {
	if c == nil || c.provider == nil {
		return AbilityDescriptorProjection{}, invalidAbilityDescriptor("AbilityDescriptor client is not initialized", nil)
	}
	return c.provider.Get(ctx, request)
}

// RuntimeAbilityDescriptorProvider reads canonical runtime descriptor catalog
// facts through the generic RuntimeAbilityClient.
type RuntimeAbilityDescriptorProvider struct {
	ability *RuntimeAbilityClient
}

func NewRuntimeAbilityDescriptorProvider(ability *RuntimeAbilityClient) (*RuntimeAbilityDescriptorProvider, error) {
	if ability == nil {
		return nil, invalidAbilityDescriptor("runtime ability client is required", nil)
	}
	return &RuntimeAbilityDescriptorProvider{ability: ability}, nil
}

func (p *RuntimeAbilityDescriptorProvider) List(ctx context.Context, request AbilityDescriptorListRequest) (AbilityDescriptorPage, error) {
	if p == nil || p.ability == nil {
		return AbilityDescriptorPage{}, invalidAbilityDescriptor("runtime AbilityDescriptor provider is not initialized", nil)
	}
	args := map[string]any{}
	if scope := strings.TrimSpace(request.Scope); scope != "" {
		args["scope"] = scope
	}
	if ownerURA := strings.TrimSpace(request.OwnerURA); ownerURA != "" {
		args["owner_ura"] = ownerURA
	}
	if abilityURA := strings.TrimSpace(request.AbilityURA); abilityURA != "" {
		args["ability_ura"] = abilityURA
	}
	output, err := p.ability.invokeCatalogueRead(ctx, request.Call, runtimeAbilityDescriptorListRoute, args)
	if err != nil {
		return AbilityDescriptorPage{}, err
	}
	rawAbilities, ok := output["abilities"].([]any)
	if !ok {
		return AbilityDescriptorPage{}, invalidAbilityDescriptor("runtime descriptor catalog output must include descriptor rows", nil)
	}
	descriptors := make([]AbilityDescriptorProjection, 0, len(rawAbilities))
	for i, raw := range rawAbilities {
		row, ok := raw.(map[string]any)
		if !ok {
			return AbilityDescriptorPage{}, invalidAbilityDescriptor(fmt.Sprintf("ability descriptor row %d must be an object", i), nil)
		}
		projection, err := projectRuntimeAbilityDescriptor(row, i)
		if err != nil {
			return AbilityDescriptorPage{}, err
		}
		descriptors = append(descriptors, projection)
	}
	return AbilityDescriptorPage{Descriptors: descriptors}, nil
}

func (p *RuntimeAbilityDescriptorProvider) Get(ctx context.Context, request AbilityDescriptorGetRequest) (AbilityDescriptorProjection, error) {
	abilityURA := strings.TrimSpace(request.AbilityURA)
	if abilityURA == "" {
		return AbilityDescriptorProjection{}, invalidAbilityDescriptor("ability_ura is required", nil)
	}
	page, err := p.List(ctx, AbilityDescriptorListRequest{
		Call:       request.Call,
		Scope:      request.Scope,
		AbilityURA: abilityURA,
	})
	if err != nil {
		return AbilityDescriptorProjection{}, err
	}
	version := strings.TrimSpace(request.DescriptorVersion)
	callMode := strings.TrimSpace(request.CallMode)
	matches := make([]AbilityDescriptorProjection, 0, len(page.Descriptors))
	for _, descriptor := range page.Descriptors {
		if descriptor.AbilityURA != abilityURA {
			return AbilityDescriptorProjection{}, invalidAbilityDescriptor("runtime returned descriptor outside requested ability_ura", nil)
		}
		if version != "" && descriptor.Version != version {
			continue
		}
		if callMode != "" && descriptor.CallMode != callMode {
			continue
		}
		matches = append(matches, descriptor)
	}
	switch len(matches) {
	case 0:
		return AbilityDescriptorProjection{}, abilityDescriptorNotFound(abilityURA)
	case 1:
		return matches[0], nil
	default:
		return AbilityDescriptorProjection{}, invalidAbilityDescriptor("ability descriptor selection is ambiguous; specify descriptor_version or call_mode", nil)
	}
}

// ProjectAbilityDescriptorRef projects a DescriptorRef through the
// product-neutral Axon-delegated Addressing seam.
func ProjectAbilityDescriptorRef(ctx context.Context, addressing Addressing, raw string) (AbilityDescriptorRef, error) {
	if addressing == nil {
		return AbilityDescriptorRef{}, invalidProfileClient(addressingProfile, "addressing provider is required for descriptor_ref projection")
	}
	projection, err := addressing.ProjectDescriptorRef(ctx, CanonicalDescriptorRefRequest{DescriptorRef: raw})
	if err != nil {
		return AbilityDescriptorRef{}, err
	}
	if projection.Kind != "descriptor_ref" ||
		strings.TrimSpace(projection.DescriptorRef) == "" ||
		strings.TrimSpace(projection.AbilityURA) == "" ||
		strings.TrimSpace(projection.DescriptorVersion) == "" {
		return AbilityDescriptorRef{}, invalidProfilePayload(addressingProfile, "descriptor_ref projection is incomplete", nil)
	}
	return AbilityDescriptorRef{
		Raw:            projection.DescriptorRef,
		AbilityURA:     projection.AbilityURA,
		Version:        projection.DescriptorVersion,
		DescriptorHash: projection.DescriptorHash,
		Action:         projection.Action,
	}, nil
}

func ProjectAbilityDescriptor(raw map[string]any) AbilityDescriptorProjection {
	projection := AbilityDescriptorProjection{
		AbilityURA:       descriptorString(raw["ability_ura"]),
		DescriptorRef:    descriptorString(raw["descriptor_ref"]),
		Name:             descriptorString(raw["name"]),
		OwnerURA:         descriptorString(raw["owner_ura"]),
		Version:          descriptorString(raw["descriptor_version"]),
		SchemaHash:       descriptorString(raw["schema_hash"]),
		DescriptorHash:   descriptorString(raw["descriptor_hash"]),
		CallMode:         descriptorString(raw["call_mode"]),
		Class:            descriptorString(raw["class"]),
		ReceiptSemantics: descriptorMap(raw["receipt_semantics"]),
		Visibility:       descriptorString(raw["visibility"]),
		Source:           descriptorString(raw["source"]),
		Description:      descriptorString(raw["description"]),
		Metadata:         descriptorMap(raw["metadata"]),
	}
	if hints := descriptorMap(raw["hints"]); hints != nil {
		projection.Hints = AbilityDescriptorHints{
			ReadOnly:      descriptorBool(hints["read_only"]),
			Destructive:   descriptorBool(hints["destructive"]),
			Idempotent:    descriptorBool(hints["idempotent"]),
			StreamingOnly: descriptorBool(hints["streaming_only"]),
			BidiOnly:      descriptorBool(hints["bidi_only"]),
		}
	}
	if schema := descriptorMap(raw["schema_summary"]); schema != nil {
		projection.SchemaSummary = schema
	}
	projection.InputSchema = descriptorMap(raw["input_schema"])
	return projection
}

func projectRuntimeAbilityDescriptor(raw map[string]any, index int) (AbilityDescriptorProjection, error) {
	projection := AbilityDescriptorProjection{
		AbilityURA:       "",
		DescriptorRef:    "",
		Name:             "",
		OwnerURA:         "",
		Version:          "",
		SchemaHash:       "",
		DescriptorHash:   "",
		CallMode:         "",
		Class:            "",
		ReceiptSemantics: nil,
		Visibility:       "",
		Source:           "",
		Description:      "",
		Metadata:         nil,
	}
	var err error
	if projection.AbilityURA, err = requiredDescriptorString(raw, "ability_ura", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.DescriptorRef, err = requiredDescriptorString(raw, "descriptor_ref", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.Name, err = requiredDescriptorString(raw, "name", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.OwnerURA, err = requiredDescriptorString(raw, "owner_ura", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.Version, err = requiredDescriptorString(raw, "descriptor_version", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.SchemaHash, err = optionalDescriptorString(raw, "schema_hash", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.DescriptorHash, err = optionalDescriptorString(raw, "descriptor_hash", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.CallMode, err = optionalDescriptorString(raw, "call_mode", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.Class, err = optionalDescriptorString(raw, "class", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.Visibility, err = optionalDescriptorString(raw, "visibility", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.Source, err = optionalDescriptorString(raw, "source", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.Description, err = optionalDescriptorString(raw, "description", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.ReceiptSemantics, err = optionalDescriptorMapWithErr(raw, "receipt_semantics", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.Metadata, err = optionalDescriptorMapWithErr(raw, "metadata", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.AbilityURA == "" ||
		projection.DescriptorRef == "" ||
		projection.Name == "" ||
		projection.OwnerURA == "" ||
		projection.Version == "" {
		return AbilityDescriptorProjection{}, invalidAbilityDescriptor(fmt.Sprintf("ability descriptor row %d is missing identity fields", index), nil)
	}
	if hints, err := optionalDescriptorMapWithErr(raw, "hints", index); err != nil {
		return AbilityDescriptorProjection{}, err
	} else if hints != nil {
		projection.Hints = AbilityDescriptorHints{
			ReadOnly:      false,
			Destructive:   false,
			Idempotent:    false,
			StreamingOnly: false,
			BidiOnly:      false,
		}
		if projection.Hints.ReadOnly, err = optionalDescriptorBool(hints, "hints.read_only", index); err != nil {
			return AbilityDescriptorProjection{}, err
		}
		if projection.Hints.Destructive, err = optionalDescriptorBool(hints, "hints.destructive", index); err != nil {
			return AbilityDescriptorProjection{}, err
		}
		if projection.Hints.Idempotent, err = optionalDescriptorBool(hints, "hints.idempotent", index); err != nil {
			return AbilityDescriptorProjection{}, err
		}
		if projection.Hints.StreamingOnly, err = optionalDescriptorBool(hints, "hints.streaming_only", index); err != nil {
			return AbilityDescriptorProjection{}, err
		}
		if projection.Hints.BidiOnly, err = optionalDescriptorBool(hints, "hints.bidi_only", index); err != nil {
			return AbilityDescriptorProjection{}, err
		}
	}
	if projection.SchemaSummary, err = optionalDescriptorMapWithErr(raw, "schema_summary", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	if projection.InputSchema, err = optionalDescriptorMapWithErr(raw, "input_schema", index); err != nil {
		return AbilityDescriptorProjection{}, err
	}
	return projection, nil
}

func descriptorString(value any) string {
	if typed, ok := value.(string); ok {
		return typed
	}
	return ""
}

func descriptorBool(value any) bool {
	typed, ok := value.(bool)
	return ok && typed
}

func descriptorMap(value any) map[string]any {
	typed, ok := value.(map[string]any)
	if ok {
		return typed
	}
	if typed, ok := value.(map[string]string); ok {
		mapped := make(map[string]any, len(typed))
		for key, raw := range typed {
			mapped[key] = raw
		}
		return mapped
	}
	return nil
}

func requiredDescriptorString(raw map[string]any, field string, index int) (string, error) {
	value, ok := raw[field]
	if !ok || value == nil {
		return "", nil
	}
	typed, ok := value.(string)
	if !ok {
		return "", invalidAbilityDescriptor(fmt.Sprintf("ability descriptor row %d field %s must be a string", index, field), nil)
	}
	return strings.TrimSpace(typed), nil
}

func optionalDescriptorString(raw map[string]any, field string, index int) (string, error) {
	value, ok := raw[field]
	if !ok || value == nil {
		return "", nil
	}
	typed, ok := value.(string)
	if !ok {
		return "", invalidAbilityDescriptor(fmt.Sprintf("ability descriptor row %d field %s must be a string", index, field), nil)
	}
	return strings.TrimSpace(typed), nil
}

func optionalDescriptorMapWithErr(raw map[string]any, field string, index int) (map[string]any, error) {
	value, ok := raw[field]
	if !ok || value == nil {
		return nil, nil
	}
	mapped := descriptorMap(value)
	if mapped == nil {
		return nil, invalidAbilityDescriptor(fmt.Sprintf("ability descriptor row %d field %s must be an object", index, field), nil)
	}
	return mapped, nil
}

func optionalDescriptorBool(raw map[string]any, field string, index int) (bool, error) {
	key := field
	if dot := strings.LastIndex(field, "."); dot >= 0 {
		key = field[dot+1:]
	}
	value, ok := raw[key]
	if !ok || value == nil {
		return false, nil
	}
	typed, ok := value.(bool)
	if !ok {
		return false, invalidAbilityDescriptor(fmt.Sprintf("ability descriptor row %d field %s must be a boolean", index, field), nil)
	}
	return typed, nil
}

func invalidAbilityDescriptor(message string, cause error) error {
	return &SDKError{
		Code:      ErrInvalidArgument,
		Stage:     "ability_descriptor",
		Retry:     RetryNever,
		Retryable: false,
		Message:   message,
		Cause:     cause,
	}
}

func abilityDescriptorNotFound(abilityURA string) error {
	return &SDKError{
		Code:      ErrDescriptorNotFound,
		Stage:     "ability_descriptor",
		Retry:     RetryNever,
		Retryable: false,
		Message:   "ability descriptor not found",
		Details:   map[string]any{"ability_ura": abilityURA},
	}
}
